infra/relayer/src/purse.js

```diff
@@ -0,0 +1,696 @@
+   * @returns {string} The transaction hash of the sent transaction
+   */
+  async sendTx(tx, onReceipt) {
+    const address = await this.getAvailableAccount()
+
+    // Set the from and nonce for the account
+    tx = {
+      ...tx,
+      from: address,
+      nonce: await this.txCount(address)
+    }
+
+    if (!tx.gasPrice) {
+      tx.gasPrice = await this.gasPrice()
+    }
+
+    logger.debug('Sending tx: ', tx)
+
+    const signed = await this.signTx(address, tx)
+    const rawTx = signed.rawTransaction
+    // Not in this version of web3.js 1.0...
+    // const txHash = signed.transactionHash
+    const txHash = this.web3.utils.sha3(rawTx)
+
+    // If there's an onReceipt cb, store it for later
+    if (onReceipt) {
+      if (typeof onReceipt !== 'function') {
+        throw new Error('onReceipt is not a function')
+      }
+      this.receiptCallbacks[txHash] = onReceipt
+    }
+
+    // In case it needs to be rebroadcast
+    this.pendingTransactions[txHash] = rawTx
+
+    // blast it
+    await sendRawTransaction(this.web3, rawTx)
+
+    await this.incrementTxCount(address)
+
+    logger.info(`Sent ${txHash}`)
+
+    return txHash
+  }
+
+  /**
+   * Get the current transaction count for an account (includes known pending)
+   * @param address {string} - The account address to increment
+   */
+  async txCount(address) {
+    let txCount = this.accounts[address] ? this.accounts[address].txCount : 0
+
+    // If we have one in memory, go with that
+    if (txCount > 0) return txCount
+
+    // Check redis, if available
+    if (txCount === 0 && this.rclient && this.rclient.connected) {
+      const countFromRedis = await this.rclient.getAsync(
+        `${REDIS_TX_COUNT_PREFIX}${address}`
+      )
+      // null defense
+      if (countFromRedis) {
+        txCount = countFromRedis
+      }
+    } else {
+      logger.warn('Redis unavailable')
+    }
+
+    /**
+     * Finally, check with the JSON-RPC provider.  Chance of some pendings screwing this up if we
+     * get this far.  I'm at least sure that eth_pendingTransactions, is completely  not supported
+     * by Alchemy.  Doubtful from Infura as well.  And it's a poor method anyway with no arguments.
+     *
+     * EDGE CASE WARNING
+     * -----------------
+     *
+     * The only time this should be an issue is if the service is magically restarted, Redis is
+     * wiped or unavailable, all while there are transactions still in the pool waiting to be
+     * mined.
+     *
+     * Also, tiny chance of transaction replacement here if the gas price goes up during the above
+     * edge case.  That would be... like a dropped transaction in the client.  Perhaps not a deal
+     * breaker, but maybe some thought should be put into that in the future as volume goes up. The
+     * bigger the load, the likelier this becomes.
+     */
+    const w3txCount = parseInt(await this.web3.eth.getTransactionCount(address))
+    if (txCount < w3txCount) txCount = w3txCount
```

Comment:
I'd log a warning that w3count was used rather than Redis. This indicates something is fishy... like at some point perhaps Redis count was not updated...

Reply:
This isn't necessarily fishy.  This was defensive, and it will be triggered at a very minimum on the initial process startup.  I'm adding the warning under the condition of `txCount > 0`.